File: client-scalaz/src/main/scala/org/openapitools/client/api/Trip.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import Trip._

case class Trip (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
owner: Option[Account],
pickupLocation: Option[Location],
pickupTime: Option[OffsetDateTime],
dropOffLocation: Option[Location],
dropOffTime: Option[OffsetDateTime],
shipments: Option[List[Shipment]],
recurrence: Option[Recurrence],
timeZone: Option[String],
route: Option[Route],
preference: Option[Preference],
notifications: Option[String],
matches: Option[List[Trip]],
ownerConnection: Option[Connection],
originalId: Option[Long])

object Trip {
  import DateTimeCodecs._
  sealed trait Preference
  case object DRIVE extends Preference
  case object FLEXIBLE extends Preference
  case object RIDE extends Preference

  object Preference {
    def toPreference(s: String): Option[Preference] = s match {
      case "DRIVE" => Some(DRIVE)
      case "FLEXIBLE" => Some(FLEXIBLE)
      case "RIDE" => Some(RIDE)
      case _ => None
    }

    def fromPreference(x: Preference): String = x match {
      case DRIVE => "DRIVE"
      case FLEXIBLE => "FLEXIBLE"
      case RIDE => "RIDE"
    }
  }

  implicit val PreferenceEnumEncoder: EncodeJson[Preference] =
    EncodeJson[Preference](is => StringEncodeJson(Preference.fromPreference(is)))

  implicit val PreferenceEnumDecoder: DecodeJson[Preference] =
    DecodeJson.optionDecoder[Preference](n => n.string.flatMap(jStr => Preference.toPreference(jStr)), "Preference failed to de-serialize")

  implicit val TripCodecJson: CodecJson[Trip] = CodecJson.derive[Trip]
  implicit val TripDecoder: EntityDecoder[Trip] = jsonOf[Trip]
  implicit val TripEncoder: EntityEncoder[Trip] = jsonEncoderOf[Trip]
}

File: client-clojure/src/sirqul_io_t_platform/specs/trip.clj
(ns sirqul-io-t-platform.specs.trip
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.location :refer :all]
            [sirqul-io-t-platform.specs.location :refer :all]
            [sirqul-io-t-platform.specs.shipment :refer :all]
            [sirqul-io-t-platform.specs.recurrence :refer :all]
            [sirqul-io-t-platform.specs.route :refer :all]
            [sirqul-io-t-platform.specs.trip :refer :all]
            [sirqul-io-t-platform.specs.connection :refer :all]
            )
  (:import (java.io File)))


(def trip-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :owner) account-spec
   (ds/opt :pickupLocation) location-spec
   (ds/opt :pickupTime) inst?
   (ds/opt :dropOffLocation) location-spec
   (ds/opt :dropOffTime) inst?
   (ds/opt :shipments) (s/coll-of shipment-spec)
   (ds/opt :recurrence) recurrence-spec
   (ds/opt :timeZone) string?
   (ds/opt :route) route-spec
   (ds/opt :preference) string?
   (ds/opt :notifications) string?
   (ds/opt :matches) (s/coll-of trip-spec)
   (ds/opt :ownerConnection) connection-spec
   (ds/opt :originalId) int?
   })

(def trip-spec
  (ds/spec
    {:name ::trip
     :spec trip-data}))
